proxylib/memcached/parser.go

```diff
@@ -0,0 +1,487 @@
+	}
+	var rules []proxylib.L7NetworkPolicyRule
+	for _, l7Rule := range l7Rules.GetL7Rules() {
+		var br MemcacheRule
+		for k, v := range l7Rule.Rule {
+			switch k {
+			case "command":
+				br.command = v
+				br.commands = MemcacheOpCodeMap[v]
+			case "keyExact":
+				br.keyExact = v
+				br.keyExactBytes = []byte(v)
+			case "keyPrefix":
+				br.keyPrefix = v
+				br.keyPrefixBytes = []byte(v)
+			case "keyRegex":
+				br.keyRegex = v
+				br.regex = regexp.MustCompile(v)
+			default:
+				proxylib.ParseError(fmt.Sprintf("Unsupported key: %s", k), rule)
+			}
+		}
+		if br.command == "" {
+			proxylib.ParseError("command not specified", rule)
+		}
+		log.Debugf("Parsed MemcacheRule pair: %v", br)
+		rules = append(rules, &br)
+	}
+	return rules
+}
+
+// MemcacheParserFactory implements proxylib.ParserFactory
+type MemcacheParserFactory struct{}
+
+// Create creates memcached parser
+func (p *MemcacheParserFactory) Create(connection *proxylib.Connection) proxylib.Parser {
+	log.Debugf("MemcacheParserFactory: Create: %v", connection)
+	return &MemcacheParser{
+		connection:   connection,
+		textParser:   text.TextMemcacheParserFactoryInstance.Create(connection),
+		binaryParser: binary.BinaryMemcacheParserFactoryInstance.Create(connection),
+	}
+}
+
+// compile time check for interface implementation
+var _ proxylib.ParserFactory = &MemcacheParserFactory{}
+
+var memcacheParserFactory *MemcacheParserFactory
+
+const (
+	parserName = "memcache"
+)
+
+func init() {
+	log.Info("init(): Registering memcacheParserFactory")
+	proxylib.RegisterParserFactory(parserName, memcacheParserFactory)
+	proxylib.RegisterL7RuleParser(parserName, L7MemcacheRuleParser)
+}
+
+// MemcacheParser implements proxylib.Parser
+type MemcacheParser struct {
+	connection   *proxylib.Connection
+	textParser   proxylib.Parser
+	binaryParser proxylib.Parser
+	isBinary     *bool
+}
+
+var _ proxylib.Parser = &MemcacheParser{}
+var t = true
+var f = false
+
+// OnData parses binary memcached data
+func (p *MemcacheParser) OnData(reply, endStream bool, dataBuffers [][]byte, offset int) (proxylib.OpType, int) {
+	if p.isBinary == nil {
+		remainingOffset := offset
+		var magicByte byte
+
+		for _, buf := range dataBuffers {
+			if len(buf) > remainingOffset {
+				magicByte = buf[remainingOffset]
+				break
+			} else {
+				remainingOffset = remainingOffset - len(buf)
+			}
+		}
+		if magicByte >= 128 {
+			p.isBinary = &t
+
+		} else {
+			p.isBinary = &f
+		}
+	}
+	if *p.isBinary {
+		return p.binaryParser.OnData(reply, endStream, dataBuffers, offset)
+	}
+	return p.textParser.OnData(reply, endStream, dataBuffers, offset)
+}
+
+type memcacheCommandSet struct {
+	text   map[string]struct{}
+	binary map[byte]struct{}
+}
+
+// empty var for filling map below which will be used as set
+var e struct{} = struct{}{}
+
+// MemcacheOpCodeMap maps human-readable names of memcached operations and groups to opcodes
+var MemcacheOpCodeMap = map[string]memcacheCommandSet{
```

Comment:
A link for the documentation of the memcache spec would help.

Also, how is this validated to be correct?

Reply:
Link added. How would you validate this?